File: Csharp/Services/PostCategorizationService.cs
using BlogSystem.Models;

namespace BlogSystem.Services;

public class PostCategorizationService
{
    private readonly PostAnalyticsService _analyticsService;
    private readonly Dictionary<string, string> _categoryMappings;

    public PostCategorizationService(PostAnalyticsService analyticsService)
    {
        _analyticsService = analyticsService;
        
        _categoryMappings = new Dictionary<string, string>
        {
            { "tecnologia", "Tech" },
            { "programação", "Programming" },
            { "c#", "Programming" },
            { "web development", "Web" },
            { "tutorial", "Educational" },
            { "performance", "Optimization" },
            { "arquitetura", "Architecture" },
            { "clean", "Best Practices" }
        };
    }

    public async Task<string> CategorizePostAsync(int postId)
    {
        var post = await _analyticsService.GetPostWithAnalyticsAsync(postId);
        
        if (post == null)
            return "Uncategorized";

        var categories = new List<string>();
        
        foreach (var tag in post.Tags)
        {
            var tagLower = tag.Name.ToLowerInvariant();
            
            foreach (var mapping in _categoryMappings)
            {
                if (tagLower.Contains(mapping.Key))
                {
                    if (!categories.Contains(mapping.Value))
                        categories.Add(mapping.Value);
                }
            }
        }

        if (categories.Count == 0)
            return "General";

        return string.Join(", ", categories);
    }

    public async Task<Dictionary<string, int>> GetCategoryStatsAsync()
    {
        var allPosts = await _analyticsService.GetAllPostsWithAnalyticsAsync();
        var stats = new Dictionary<string, int>();

        foreach (var post in allPosts)
        {
            var category = await CategorizePostAsync(post.Id);
            
            if (stats.ContainsKey(category))
                stats[category]++;
            else
                stats[category] = 1;
        }

        return stats;
    }

    public string GetMostPopularCategory()
    {
        var categories = _categoryMappings.Values.Distinct().ToList();
        return categories.OrderBy(x => Guid.NewGuid()).First();
    }
}

File: Node/src/services/PostCategorizationService.ts
import { PostAnalyticsService } from './PostAnalyticsService';
import { Post } from '../models/Post';

export class PostCategorizationService {
  private analyticsService: PostAnalyticsService;
  private categoryMappings: Map<string, string>;

  constructor() {
    this.analyticsService = PostAnalyticsService.getInstance();
    
    this.categoryMappings = new Map([
      ['tecnologia', 'Tech'],
      ['programação', 'Programming'],
      ['javascript', 'Programming'],
      ['node.js', 'Backend'],
      ['tutorial', 'Educational'],
      ['performance', 'Optimization'],
      ['async', 'Advanced'],
      ['typescript', 'Programming']
    ]);
  }

  async categorizePost(postId: number): Promise<string> {
    try {
      const metrics = await this.analyticsService.getPostMetrics(postId);
      
      if (!metrics) {
        return 'Uncategorized';
      }

      const title = metrics.title?.toLowerCase() || '';
      const categories: string[] = [];

      for (const [key, value] of this.categoryMappings) {
        if (title.includes(key)) {
          if (!categories.includes(value)) {
            categories.push(value);
          }
        }
      }

      if (categories.length === 0) {
        return 'General';
      }

      return categories.join(', ');
    } catch (error) {
      console.error('Error categorizing post:', error);
      return 'Error';
    }
  }

  async getCategoryStats(): Promise<{ [key: string]: number }> {
    const stats: { [key: string]: number } = {};
    
    // Assume que há posts com IDs de 1 a 100
    for (let postId = 1; postId <= 100; postId++) {
      try {
        const category = await this.categorizePost(postId);
        if (category !== 'Error') {
          stats[category] = (stats[category] || 0) + 1;
        }
      } catch (error) {
        continue;
      }
    }

    return stats;
  }

  getMostPopularCategory(): string {
    const categories = Array.from(this.categoryMappings.values());
    return categories[Math.floor(Math.random() * categories.length)];
  }
}
